File: ch2/2.2i_big_integer/primaryarithmetic/primaryarithmetic.py
import sys

def solve(a, b):
    i = 0
    res = 0
    carry = 0
    while (i < len(a) and i < len(b)):
        sum = a[i] + b[i] + carry
        carry = int(sum / 10)
        if (carry > 0): res += 1
        i += 1

    while (i < len(a)):
        sum = a[i] + carry
        carry = int(sum / 10)
        if (carry > 0): res += 1
        i += 1

    while (i < len(b)):
        sum = b[i] + carry
        carry = int(sum / 10)
        if (carry > 0): res += 1
        i += 1
    
    return res


for line in sys.stdin.readlines():
    [a, b] = line.strip().split(' ')
    if (a == '0' and b == '0'): break
    a = [int(char) for char in a]
    a.reverse()
    b = [int(char) for char in b]
    b.reverse()

    res = solve(a, b)

    if (res == 0):
        print("No carry operation.")
    elif (res == 1):
        print("1 carry operation.")
    else:
        print(res, "carry operations.")

File: kattis/primaryarithmetic.cpp
#include <bits/stdc++.h>
#define ll long long
#define ii pair<ll, ll>
#define iii tuple<ll, ll, ll>
using namespace std;

// TODO: this solution is not fully correct yet.

ll solve(ll a, ll b) {
    
    ll d = 10;
    ll carry = 0;

    ll res = 0;
    while (a > 0 && b > 0) {
        ll a1 = (a - (a / d * d)) / (d/10);
        ll b1 = (b - (b / d * d)) / (d/10);

        // printf("%d %d %d %d %d %d\n", d, a, b, a1, b1, carry);

        ll sum = a1 + b1 + carry;
        carry = sum / 10;
        if (carry > 0) res++;

        a -= a1 * (d/10);
        b -= b1 * (d/10);
        d *= 10;
    }

    while (a > 0) {
        ll a1 = (a - (a / d * d)) / (d/10);
        ll sum = a1 + carry;
        carry = sum / 10;
        if (carry > 0) res++;
        a -= a1 * (d/10);
        d *= 10;
    }

    while (b > 0) {
        ll b1 = (b - (b / d * d)) / (d/10);
        ll sum = b1 + carry;
        carry = sum / 10;
        if (carry > 0) res++;
        b -= b1 * (d/10);
        d *= 10;
    }

    while (carry > 0) {
        carry = carry / 10;
        if (carry > 0) res++;
    }

    return res;
}

int main() {

    ll a,b;

    while (cin>>a>>b, a&&b ) {
        ll res = solve(a, b);
        if (res == 0) {
            cout << "No carry operation.\n";
        } else if (res == 1) {
            cout << "1 carry operation.\n";
        } else {
            cout << res << " carry operations.\n";
        }
    }

}
